File: eMailManager2010/Controls/FileNameFilterBox.vb
Imports System.ComponentModel

Public Class FileNameFilterBox
    Implements INotifyPropertyChanged

    Private m_FileNameFilter As String
    Private m_DefaultFileNameFilter As String = "<sent>_<from>_<subject>"
    Private m_IsValid As Boolean = True

    Public Property FileNameFilter() As String
        Get
            Return m_FileNameFilter
        End Get
        Set(ByVal value As String)
            m_FileNameFilter = value
            ' Call OnPropertyChanged whenever the property is updated
            OnPropertyChanged("FileNameFilter")
        End Set
    End Property
    Public Property DefaultFileNameFilter() As String
        Get
            Return m_DefaultFileNameFilter
        End Get
        Set(ByVal value As String)
            m_DefaultFileNameFilter = value
        End Set
    End Property
    Public ReadOnly Property IsValid() As Boolean
        Get
            Return m_IsValid
        End Get
    End Property

    Public Event IsValidChanged()
    Public Event PropertyChanged As PropertyChangedEventHandler Implements INotifyPropertyChanged.PropertyChanged

    ' Create the OnPropertyChanged method to raise the event
    Protected Sub OnPropertyChanged(ByVal name As String)
        RaiseEvent PropertyChanged(Me, New PropertyChangedEventArgs(name))

        Select Case name
            Case "FileNameFilter"
                Me.TextBox1.Text = m_FileNameFilter
        End Select

    End Sub

    Private Sub ComboBox1_SelectionChangeCommitted(ByVal sender As Object, ByVal e As System.EventArgs) Handles ComboBox1.SelectionChangeCommitted
        Me.TextBox1.Text = Me.TextBox1.Text & Me.ComboBox1.SelectedItem.ToString
        m_FileNameFilter = Me.TextBox1.Text

        Me.TextBox1.Focus()
        Me.TextBox1.Select(Me.TextBox1.Text.Length, Me.TextBox1.Text.Length)
    End Sub

    Private Sub TextBox1_TextChanged(ByVal sender As Object, ByVal e As System.EventArgs) Handles TextBox1.TextChanged
        If Me.TextBox1.Text = String.Empty Then
            m_IsValid = False
            Me.TextBox1.BackColor = Drawing.Color.Red
            'Me.TextBox1.Text = m_DefaultFileNameFilter
            RaiseEvent IsValidChanged()
        Else
            m_IsValid = True
            Me.TextBox1.BackColor = Drawing.Color.White
            RaiseEvent IsValidChanged()
        End If

        m_FileNameFilter = Me.TextBox1.Text
    End Sub
End Class

File: eMailManager/Controls/FileNameFilterBox.cs
using System;
using System.ComponentModel;
using System.Windows.Forms;

namespace eMailManager.Controls
{
    public partial class FileNameFilterBox : INotifyPropertyChanged
    {
        public FileNameFilterBox()
        {
            InitializeComponent();
            _TextBox1.Name = "TextBox1";
            _ComboBox1.Name = "ComboBox1";
        }

        private string _fileNameFilter;
        private string _defaultFileNameFilter = "<sent>_<from>_<subject>";
        private bool _isValid = true;

        public string FileNameFilter
        {
            get
            {
                return _fileNameFilter;
            }

            set
            {
                _fileNameFilter = value;
                // Call OnPropertyChanged whenever the property is updated
                OnPropertyChanged("FileNameFilter");
            }
        }

        public string DefaultFileNameFilter
        {
            get
            {
                return _defaultFileNameFilter;
            }

            set
            {
                _defaultFileNameFilter = value;
            }
        }

        public bool IsValid
        {
            get
            {
                return _isValid;
            }
        }

        public event IsValidChangedEventHandler IsValidChanged;

        public delegate void IsValidChangedEventHandler();

        public event PropertyChangedEventHandler PropertyChanged;

        // Create the OnPropertyChanged method to raise the event
        protected void OnPropertyChanged(string name)
        {
            PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(name));
            switch (name ?? "")
            {
                case "FileNameFilter":
                {
                    TextBox1.Text = _fileNameFilter;
                    break;
                }
            }
        }

        private void ComboBox1_SelectionChangeCommitted(object sender, EventArgs e)
        {
            TextBox1.Text = TextBox1.Text + ComboBox1.SelectedItem.ToString();
            _fileNameFilter = TextBox1.Text;
            TextBox1.Focus();
            TextBox1.Select(TextBox1.Text.Length, TextBox1.Text.Length);
        }

        private void TextBox1_TextChanged(object sender, EventArgs e)
        {
            if ((TextBox1.Text ?? "") == (string.Empty ?? ""))
            {
                _isValid = false;
                TextBox1.BackColor = System.Drawing.Color.Red;
                TextBox1.Text = _defaultFileNameFilter;
                IsValidChanged?.Invoke();
            }
            else
            {
                _isValid = true;
                TextBox1.BackColor = System.Drawing.Color.White;
                IsValidChanged?.Invoke();
            }

            _fileNameFilter = TextBox1.Text;
        }
    }
}